In [1]:
import matplotlib.pyplot as plt
import pandas as pd
import os

In [2]:
# Set option to display all columns
pd.set_option('display.max_columns', None)

In [3]:
# Get the current script's directory
current_script_dir = os.getcwd()

# Move up to the parent directory
parent_dir = os.path.dirname(current_script_dir)

# Define the sibling data directory name
data_dir = os.path.join(parent_dir, 'data')

# Viewing the attached feature documentation of the dataset

In [4]:
# Read the documentation of the dataset
doc_filename = 'documentation.csv'
doc_file_path = os.path.join(data_dir, doc_filename)
doc_df = pd.read_csv(doc_file_path)

In [5]:
pd.set_option('display.max_colwidth', None)  # Show all text in cells
pd.set_option('display.max_rows', None)

display(doc_df)

pd.reset_option('display.max_rows')
pd.reset_option('display.max_colwidth')

SYS_FIELD_NAME  \
0                    YEAR   
1                 QUARTER   
2                   MONTH   
3            DAY_OF_MONTH   
4             DAY_OF_WEEK   
5                 FL_DATE   
6       OP_UNIQUE_CARRIER   
7   OP_CARRIER_AIRLINE_ID   
8              OP_CARRIER   
9                TAIL_NUM   
10      OP_CARRIER_FL_NUM   
11      ORIGIN_AIRPORT_ID   
12  ORIGIN_AIRPORT_SEQ_ID   
13  ORIGIN_CITY_MARKET_ID   
14                 ORIGIN   
15       ORIGIN_CITY_NAME   
16       ORIGIN_STATE_ABR   
17      ORIGIN_STATE_FIPS   
18        ORIGIN_STATE_NM   
19             ORIGIN_WAC   
20        DEST_AIRPORT_ID   
21    DEST_AIRPORT_SEQ_ID   
22    DEST_CITY_MARKET_ID   
23                   DEST   
24         DEST_CITY_NAME   
25         DEST_STATE_ABR   
26        DEST_STATE_FIPS   
27          DEST_STATE_NM   
28               DEST_WAC   
29           CRS_DEP_TIME   
30               DEP_TIME   
31              DEP_DELAY   
32          DEP_DELAY_NEW   
33              DEP_DEL15   
34        DEP_DELAY_GROUP   
35           DEP_TIME_BLK   
36               TAXI_OUT   
37             WHEELS_OFF   
38              WHEELS_ON   
39                TAXI_IN   
40           CRS_ARR_TIME   
41               ARR_TIME   
42              ARR_DELAY   
43          ARR_DELAY_NEW   
44              ARR_DEL15   
45        ARR_DELAY_GROUP   
46           ARR_TIME_BLK   
47              CANCELLED   
48      CANCELLATION_CODE   
49               DIVERTED   
50       CRS_ELAPSED_TIME   
51    ACTUAL_ELAPSED_TIME   
52               AIR_TIME   
53                FLIGHTS   
54               DISTANCE   
55         DISTANCE_GROUP   
56          CARRIER_DELAY   
57          WEATHER_DELAY   
58              NAS_DELAY   
59         SECURITY_DELAY   
60    LATE_AIRCRAFT_DELAY   

                                                                                                                                                                                                                                                 FIELD_DESC  
0                                                                                                                                                                                                                                                      Year  
1                                                                                                                                                                                                                                             Quarter (1-4)  
2                                                                                                                                                                                                                                                     Month  
3                                                                                                                                                                                                                                              Day of Month  
4                                                                                                                                                                                                                                               Day of Week  
5                                                                                                                                                                                                                                    Flight Date (yyyymmdd)  
6                                               Unique Carrier Code. When the same code has been used by multiple carriers, a numeric suffix is used for earlier users, for example, PA, PA(1), PA(2). Use this field for analysis across a range of years.  
7       An identification number assigned by US DOT to identify a unique airline (carrier). A unique airline (carrier) is defined as one holding and reporting under the same DOT ce

## What is useful for the goal of predicting deparature delays?
In short, we need information we know before the actual deparature of a future flight so we can use such information to make predictions. Thus we need information about:

- Time but not year since we are only using 2022
- Carrier ID, Flight Number, and tail number
- Scheduled deparature and arrival times
- Origin Airport, city and state
- Destination Airport, city and state
- Distance and Scheduled air time between origin and destination
- Historic Delay information that is out target to predict
- Number of scheduled flights between origin and destination
  
## Useful Features
1. QUARTER
2. MONTH
3. DAY_OF_MONTH
4. DAY_OF_WEEK
5. OP_UNIQUE_CARRIER
6. OP_CARRIER_AIRLINE_ID
7. TAIL_NUM
8. OP_CARRIER_FL_NUM
9. ORIGIN_AIRPORT_ID 
10. ORIGIN_CITY_MARKET_ID
11. ORIGIN_STATE_FIPS
12. ORIGIN_WAC
13. DEST_AIRPORT_ID
14. DEST_CITY_MARKET_ID
15. DEST_STATE_FIPS
16. DEST_WAC
17. CRS_DEP_TIME
18. DEP_DELAY
19. DEP_DELAY_NEW
20. DEP_DEL15
21. DEP_DELAY_GROUP
22. DEP_TIME_BLK
23. CRS_ARR_TIME
24. ARR_TIME_BLK
25. CRS_ELAPSED_TIME
26. DISTANCE
27. DISTANCE_GROUP
28. CARRIER_DELAY
29. WEATHER_DELAY
30. NAS_DELAY
31. SECURITY_DELAY
32. LATE_AIRCRAFT_DELAY
33. FLIGHTS

## Non-usefull Features
1. Year
2. FL_DATE (Month and day are already separated in different features)
3. OP_CARRIER (The same code may have been assigned to different carriers over time, we need to be able to 
distinguish between carriers thus better to use OP_UNIQUE_CARRIER)
4. ORIGIN_AIRPORT_SEQ_ID (It is only unique at a given point of time and can change for an airport in the future)
5. ORIGIN (No need since ORIGIN_AIRPORT_ID is available)
6. ORIGIN_CITY_NAME (No need since ORIGIN_CITY_MARKET_ID is available)
7. ORIGIN_STATE_ABR (It consists of 2 letters and thus categorical, better to have a numerical value for training of ML models)
8. ORIGIN_STATE_NM (No need since ORIGIN_STATE_FIPS is available)
9. DEST_AIRPORT_SEQ_ID (It is only unique at a given point of time and can change for an airport in the future)
10. DEST (No need since DEST_AIRPORT_ID is available)
11. DEST_CITY_NAME (No need since DEST_CITY_MARKET_ID is available)
12. DEST_STATE_ABR (It consists of 2 letters and thus categorical, better to have a numerical value for training of ML models)
13. DEST_STATE_NM (No need since DEST_STATE_FIPS is available)
14. DEP_TIME (No need for this field as we already given the delay and groups in other columns such as DEP_DELAY)
15. TAXI_OUT (Actual and not schedualed, we dont no this when we need to make a prediction)
16. WHEELS_OFF (Actual and not schedualed, we dont no this when we need to make a prediction)
17. WHEELS_ON (Actual and not schedualed, we dont no this when we need to make a prediction)
18. TAXI_IN (Actual and not schedualed, we dont no this when we need to make a prediction)
19. ARR_TIME (Actual and not schedualed, we dont no this when we need to make a prediction)
20. ARR_DELAY (We are interested in deparature delay only)
21. ARR_DELAY_NEW (We are interested in deparature delay only)
22. ARR_DEL15 (We are interested in deparature delay only)
23. ARR_DELAY_GROUP (We are interested in deparature delay only)
24. ACTUAL_ELAPSED_TIME (Actual and not schedualed, we dont no this when we need to make a prediction)
25. AIR_TIME (Actual and not schedualed, we dont no this when we need to make a prediction)
26. CANCELLED
27. CANCELLATION_CODE
28. DIVERTED

In [6]:
## Set the useful features in a columns array
columns = [
    "QUARTER",
    "MONTH",
    "DAY_OF_MONTH",
    "DAY_OF_WEEK",
    
    "OP_UNIQUE_CARRIER",
    "OP_CARRIER_AIRLINE_ID",
    "TAIL_NUM",
    "OP_CARRIER_FL_NUM",
    
    "ORIGIN_AIRPORT_ID",
    "ORIGIN_CITY_MARKET_ID",
    "ORIGIN_STATE_FIPS",
    "ORIGIN_WAC",
    "DEST_AIRPORT_ID",
    "DEST_CITY_MARKET_ID",
    "DEST_STATE_FIPS",
    "DEST_WAC",
    
    "CRS_DEP_TIME",
    "DEP_TIME_BLK",
    "CRS_ARR_TIME",
    "ARR_TIME_BLK",
    "CRS_ELAPSED_TIME",
    "DISTANCE",
    "DISTANCE_GROUP",
    "FLIGHTS",
    
    "DEP_DELAY",
    "DEP_DELAY_NEW",
    "DEP_DEL15",
    "DEP_DELAY_GROUP",
    "CARRIER_DELAY",
    "WEATHER_DELAY",
    "NAS_DELAY",
    "SECURITY_DELAY",
    "LATE_AIRCRAFT_DELAY"
]

In [7]:
# Create an empty dataFrame to store the dataset
df = pd.DataFrame(columns=columns)

# Define the year data directory name
year = "2022"
year_dir = os.path.join(data_dir, year)
    
for month in range(1, 13, 1):
    month_filename = f'{str(month).zfill(2)}_2022.csv'
        
    # Join the directory and filename
    file_path = os.path.join(year_dir, month_filename)
    
    # Read the CSV file into a DataFrame
    month_df = pd.read_csv(file_path)
    
    # Only keep useful data colums
    month_df = month_df[columns]
    
    # Concatenate the empty DataFrame with the this month data    
    if not df.empty:
        df = pd.concat([df, month_df], ignore_index=True)
    else:
        df = month_df

In [8]:
# Display the first few rows of the DataFrame
display(df)

QUARTER  MONTH  DAY_OF_MONTH  DAY_OF_WEEK OP_UNIQUE_CARRIER  \
0              1      1             1            6                9E   
1              1      1             1            6                9E   
2              1      1             1            6                9E   
3              1      1             1            6                9E   
4              1      1             1            6                9E   
...          ...    ...           ...          ...               ...   
6729120        4     12            31            6                YX   
6729121        4     12            31            6                YX   
6729122        4     12            31            6                YX   
6729123        4     12            31            6                YX   
6729124        4     12            31            6                YX   

         OP_CARRIER_AIRLINE_ID TAIL_NUM  OP_CARRIER_FL_NUM  ORIGIN_AIRPORT_ID  \
0                        20363   N138EV               4732              10135   
1                        20363   N138EV               4979              11433   
2                        20363   N138EV               5430              10135   
3                        20363   N138EV               5430              11433   
4                        20363   N146PQ               4722              11433   
...                        ...      ...                ...                ...   
6729120                  20452   N882RW               5741              12953   
6729121                  20452   N979RP               3607              14321   
6729122                  20452   N979RP               3686              13930   
6729123                  20452   N979RP               3699              11003   
6729124                  20452   N979RP               3719              13930   

         ORIGIN_CITY_MARKET_ID  ORIGIN_STATE_FIPS  ORIGIN_WAC  \
0                        30135                 42          23   
1                        31295                 26          43   
2                        30135                 42          23   
3                        31295                 26          43   
4                        31295                 26          43   
...                        ...                ...         ...   
6729120                  31703                 36          22   
6729121                  34321                 23          12   
6729122                  30977                 17          41   
6729123                  31003                 19          61   
6729124                  30977                 17          41   

         DEST_AIRPORT_ID  DEST_CITY_MARKET_ID  DEST_STATE_FIPS  DEST_WAC  \
0                  11433                31295               26        43   
1                  14524                34524               51        38   
2                  11433                31295               26        43   
3                  10135                30135               42        23   
4                  11423                31423               19        61   
...                  ...                  ...              ...       ...   
6729120            11066                31066               39        44   
6729121            13930                30977               17        41   
6729122            14321                34321               23        12   
6729123            13930                30977               17        41   
6729124            11003                31003               19        61   

         CRS_DEP_TIME DEP_TIME_BLK  CRS_ARR_TIME ARR_TIME_BLK  \
0                1015    1000-1059          1209    1200-1259   
1                2130    2100-2159          2305    2300-2359   
2                1700    1700-1759          1853    1800-1859   
3                1422    1400-1459          1548    1500-1559   
4                1255    1200-1259          1348    1300-1359   
...               ...          ...           ...          ...   
6729120           830 

## Target Data Quality Requirement
- The table below shows what requirement should be met by each target column for a satisfying data quality

|                         Target                           |                 Data Quality Requirement                        |
|----------------------------------------------------------|-----------------------------------------------------------------|
| DEP_DELAY                                                | Value is required; If Nan, delete row                           |
|                                                          | Value type is int                                               |
| DEP_DELAY_NEW                                            | Value is required; If Nan, delete row                           |
|                                                          | Value type is int                                               |
|                                                          | Negative values are invalid                                     |
| DEP_DEL15                                                | Value must be 1 or 0                                            |
|                                                          | Value is required; If Nan, delete row                           |
|                                                          | Value type is int                                               |
| DEP_DELAY_GROUP                                          | Values must range from -2 to 12, by 1 step                      |
|                                                          | Value is required; If Nan, delete row                           |
|                                                          | Value type is int                                               |
| CARRIER_DELAY                                            | Value is not required; If Nan, set to 0                         |
|                                                          | Negative values are invalid                                     |
|                                                          | Value type is int                                               |
| WEATHER_DELAY                                            | Value is not required; If Nan, set to 0                         |
|                                                          | Negative values are invalid                                     |
|                                                          | Value type is int                                               |
| NAS_DELAY                                                | Value is not required; If Nan, set to 0                         |
|                                                          | Negative values are invalid                                     |
|                                                          | Value type is int                                               |
| SECURITY_DELAY                                           | Value is not required; If Nan, set to 0                         |
|                                                          | Negative values are invalid                                     |
|                                                          | Value type is int                                               |
| LATE_AIRCRAFT_DELAY                                      | Value is not required; If Nan, set to 0                         |
|                                                          | Negative values are invalid                                     |
|                                                          | Value type is int                                               |

In [9]:
# We start by viewing the descriptiuon statstics of each target column
targets = [
    "DEP_DELAY",
    "DEP_DELAY_NEW",
    "DEP_DEL15",
    "DEP_DELAY_GROUP",
    "CARRIER_DELAY",
    "WEATHER_DELAY",
    "NAS_DELAY",
    "SECURITY_DELAY",
    "LATE_AIRCRAFT_DELAY"
]

display(df[targets].describe().round())

DEP_DELAY  DEP_DELAY_NEW  DEP_DEL15  DEP_DELAY_GROUP  CARRIER_DELAY  \
count  6551777.0      6551777.0  6551777.0        6551777.0      1376798.0   
mean        13.0           16.0        0.0              0.0           27.0   
std         53.0           52.0        0.0              2.0           74.0   
min        -96.0            0.0        0.0             -2.0            0.0   
25%         -5.0            0.0        0.0             -1.0            0.0   
50%         -2.0            0.0        0.0             -1.0            7.0   
75%         10.0           10.0        0.0              0.0           26.0   
max       3433.0         3433.0        1.0             12.0         3423.0   

       WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  LATE_AIRCRAFT_DELAY  
count      1376798.0  1376798.0       1376798.0            1376798.0  
mean             4.0       11.0             0.0                 25.0  
std             31.0       29.0             4.0                 55.0  
min              0.0        0.0             0.0                  0.0  
25%              0.0        0.0             0.0                  0.0  
50%              0.0        0.0             0.0                  0.0  
75%              0.0       15.0             0.0                 30.0  
max           2363.0     1740.0          1245.0               2302.0

In [10]:
prosessed_df = df.copy()

In [11]:
## DEP_DELAY
# Value is required; If Nan, delete row
# Value type is int

print("Removing NaN from 'DEP_DELAY'") 
# Get Nan values in target
nan_values = prosessed_df['DEP_DELAY'].isna()

# Get number of Nan values in target
nan_invalid_count = nan_values.sum()
print(f"{nan_invalid_count = }")

# Get number of valid exiting values in target
valid_count = (~nan_values).sum()
print(f"{valid_count = }")
  
# Drop Nan values
prosessed_df = prosessed_df.dropna(subset=['DEP_DELAY'])
print(f"{len(prosessed_df) = }")
print()

# Convert to int
print("Converting to dtype int")
print(f"Before {prosessed_df['DEP_DELAY'].dtype = }")
prosessed_df['DEP_DELAY'] = prosessed_df['DEP_DELAY'].astype(int)
print(f"After {prosessed_df['DEP_DELAY'].dtype = }")

Removing NaN from 'DEP_DELAY'
nan_invalid_count = np.int64(177348)
valid_count = np.int64(6551777)
len(prosessed_df) = 6551777

Converting to dtype int
Before prosessed_df['DEP_DELAY'].dtype = dtype('float64')
After prosessed_df['DEP_DELAY'].dtype = dtype('int64')


In [12]:
## DEP_DELAY_NEW
# Value is required; If Nan, delete row
# Value type is int
# Negative values are invalid

print("Removing NaN from 'DEP_DELAY_NEW'") 
# Get Nan values in target
nan_values = prosessed_df['DEP_DELAY_NEW'].isna()

# Get number of Nan values in target
nan_invalid_count = nan_values.sum()
print(f"{nan_invalid_count = }")

# Get number of valid exiting values in target
valid_count = (~nan_values).sum()
print(f"{valid_count = }")
  
# Drop Nan values
prosessed_df = prosessed_df.dropna(subset=['DEP_DELAY_NEW'])
print(f"{len(prosessed_df) = }")
print()

# Convert to int
print("Converting to dtype int")
print(f"Before {prosessed_df['DEP_DELAY_NEW'].dtype = }")
prosessed_df['DEP_DELAY_NEW'] = prosessed_df['DEP_DELAY_NEW'].astype(int)
print(f"After {prosessed_df['DEP_DELAY_NEW'].dtype = }")
print()

# Remove rows with invalid negative column at this column
print("Remove rows with invalid negative column at this column")
negative_count = (prosessed_df['DEP_DELAY_NEW'] < 0).sum()
print(f"{negative_count = }")
prosessed_df = prosessed_df[prosessed_df['DEP_DELAY_NEW'] >= 0]
print(f"{len(prosessed_df) = }")
print()

Removing NaN from 'DEP_DELAY_NEW'
nan_invalid_count = np.int64(0)
valid_count = np.int64(6551777)
len(prosessed_df) = 6551777

Converting to dtype int
Before prosessed_df['DEP_DELAY_NEW'].dtype = dtype('float64')
After prosessed_df['DEP_DELAY_NEW'].dtype = dtype('int64')

Remove rows with invalid negative column at this column
negative_count = np.int64(0)
len(prosessed_df) = 6551777



In [13]:
## DEP_DELAY15
# Value is required; If Nan, delete row
# Value type is int
# Values must range from -2 to 12, by 1 step

print("Removing NaN from 'DEP_DELAY15'") 
# Get Nan values in target
nan_values = prosessed_df['DEP_DEL15'].isna()

# Get number of Nan values in target
nan_invalid_count = nan_values.sum()
print(f"{nan_invalid_count = }")

# Get number of valid exiting values in target
valid_count = (~nan_values).sum()
print(f"{valid_count = }")
  
# Drop Nan values
prosessed_df = prosessed_df.dropna(subset=['DEP_DEL15'])
print(f"{len(prosessed_df) = }")
print()

# Convert to int
print("Converting to dtype int")
print(f"Before {prosessed_df['DEP_DEL15'].dtype = }")
prosessed_df['DEP_DEL15'] = prosessed_df['DEP_DEL15'].astype(int)
print(f"After {prosessed_df['DEP_DEL15'].dtype = }")
print()

# Drop rows where values in column 'DEP_DEL15' are not in [0, 1]
print("Dropping rows where values in column 'DEP_DEL15' are not in [0, 1]")
non_valid_count = (~prosessed_df['DEP_DEL15'].isin([0, 1])).sum()
print(f"{non_valid_count = }")
prosessed_df = prosessed_df[prosessed_df['DEP_DEL15'].isin([0, 1])]
print(f"After {len(prosessed_df) = }")

Removing NaN from 'DEP_DELAY15'
nan_invalid_count = np.int64(0)
valid_count = np.int64(6551777)
len(prosessed_df) = 6551777

Converting to dtype int
Before prosessed_df['DEP_DEL15'].dtype = dtype('float64')
After prosessed_df['DEP_DEL15'].dtype = dtype('int64')

Dropping rows where values in column 'DEP_DEL15' are not in [0, 1]
non_valid_count = np.int64(0)
After len(prosessed_df) = 6551777


In [14]:
## DEP_DELAY_GROUP
# Value is required; If Nan, delete row
# Value type is int
# Values must range from -2 to 12, by 1 step

print("Removing NaN from 'DEP_DELAY_GROUP'") 
# Get Nan values in target
nan_values = prosessed_df['DEP_DELAY_GROUP'].isna()

# Get number of Nan values in target
nan_invalid_count = nan_values.sum()
print(f"{nan_invalid_count = }")

# Get number of valid exiting values in target
valid_count = (~nan_values).sum()
print(f"{valid_count = }")
  
# Drop Nan values
prosessed_df = prosessed_df.dropna(subset=['DEP_DELAY_GROUP'])
print(f"{len(prosessed_df) = }")
print()

# Convert to int
print("Converting to dtype int")
print(f"Before {prosessed_df['DEP_DELAY_GROUP'].dtype = }")
prosessed_df['DEP_DELAY_GROUP'] = prosessed_df['DEP_DELAY_GROUP'].astype(int)
print(f"After {prosessed_df['DEP_DELAY_GROUP'].dtype = }")
print()

# Drop rows where values in column 'DEP_DEL15' are not in [-2, -1, 0, ..., 10, 11, 12]
print("Dropping rows where values in column 'DEP_DELAY_GROUP' are not in [-2, -1, 0, ..., 10, 11, 12]")
non_valid_count = (~prosessed_df['DEP_DELAY_GROUP'].isin(list(range(-2, 13, 1)))).sum()
print(f"{non_valid_count = }")
prosessed_df = prosessed_df[prosessed_df['DEP_DELAY_GROUP'].isin(list(range(-2, 13, 1)))]
print(f"After {len(prosessed_df) = }")

Removing NaN from 'DEP_DELAY_GROUP'
nan_invalid_count = np.int64(0)
valid_count = np.int64(6551777)
len(prosessed_df) = 6551777

Converting to dtype int
Before prosessed_df['DEP_DELAY_GROUP'].dtype = dtype('float64')
After prosessed_df['DEP_DELAY_GROUP'].dtype = dtype('int64')

Dropping rows where values in column 'DEP_DELAY_GROUP' are not in [-2, -1, 0, ..., 10, 11, 12]
non_valid_count = np.int64(0)
After len(prosessed_df) = 6551777


In [15]:
## CARRIER_DELAY, WEATHER_DELAY, NAS_DELAY, SECURITY_DELAY, LATE_AIRCRAFT_DELAY
# All of the 5 above columns have the same following requirements
# Value is required; If Nan, set to 0
# Value type is int
# Negative values are invalid, delete row

for column in ["CARRIER_DELAY", "WEATHER_DELAY", "NAS_DELAY", "SECURITY_DELAY", "LATE_AIRCRAFT_DELAY"]:
    print(f"Filling NaN with 0 from '{column}'")
    before_zero_count = (prosessed_df[column] == 0).sum()
    print(f"{before_zero_count = }")
    # Get Nan values in target
    nan_values = prosessed_df[column].isna()
    # Get number of Nan values in target
    nan_count = nan_values.sum()
    print(f"{nan_count = }")
    # Fill with 0
    prosessed_df[column] = prosessed_df[column].fillna(0)
    # Get 0 count
    zero_count = (prosessed_df[column] == 0).sum()
    print(f"{zero_count = } should be equal to {before_zero_count+nan_count}")
    print()
    
    # Convert to int
    print("Converting to dtype int")
    print(f"Before {prosessed_df[column].dtype = }")
    prosessed_df[column] = prosessed_df[column].astype(int)
    print(f"After {prosessed_df[column].dtype = }")
    print()

    # Delete rows with invalid negative values
    print(f"Dropping rows with invalid negative values in {column}")
    # Count the number of negative values in column
    negative_count = (prosessed_df[column] < 0).sum()
    print(f"{negative_count = }")
    # Drop rows where column 'A' has negative values
    prosessed_df = prosessed_df[prosessed_df[column] >= 0]
    print(f"{len(prosessed_df) = }")
    print("----------------------------------------------------------------------------------------")
    print()


Filling NaN with 0 from 'CARRIER_DELAY'
before_zero_count = np.int64(538461)
nan_count = np.int64(5174979)
zero_count = np.int64(5713440) should be equal to 5713440

Converting to dtype int
Before prosessed_df[column].dtype = dtype('float64')
After prosessed_df[column].dtype = dtype('int64')

Dropping rows with invalid negative values in CARRIER_DELAY
negative_count = np.int64(0)
len(prosessed_df) = 6551777
----------------------------------------------------------------------------------------

Filling NaN with 0 from 'WEATHER_DELAY'
before_zero_count = np.int64(1300954)
nan_count = np.int64(5174979)
zero_count = np.int64(6475933) should be equal to 6475933

Converting to dtype int
Before prosessed_df[column].dtype = dtype('float64')
After prosessed_df[column].dtype = dtype('int64')

Dropping rows with invalid negative values in WEATHER_DELAY
negative_count = np.int64(0)
len(prosessed_df) = 6551777
---------------------------------------------------------------------------------------

In [16]:
display(prosessed_df[targets])

DEP_DELAY  DEP_DELAY_NEW  DEP_DEL15  DEP_DELAY_GROUP  CARRIER_DELAY  \
0               -1              0          0               -1              0   
3               -3              0          0               -1              0   
6               -5              0          0               -1              0   
7               -3              0          0               -1              0   
8               -3              0          0               -1              0   
...            ...            ...        ...              ...            ...   
6729120         -7              0          0               -1              0   
6729121         -7              0          0               -1              0   
6729122         -7              0          0               -1              0   
6729123        -10              0          0               -1              0   
6729124         -6              0          0               -1              0   

         WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  LATE_AIRCRAFT_DELAY  
0                    0          0               0                    0  
3                    0          0               0                    0  
6                    0          0               0                    0  
7                    0          0               0                    0  
8                    0          0               0                    0  
...                ...        ...             ...                  ...  
6729120              0          0               0                    0  
6729121              0         25               0                    0  
6729122              0          0               0                    0  
6729123              0          0               0                    0  
6729124              0          0               0                    0  

[6551777 rows x 9 columns]

In [17]:
display(prosessed_df[targets].describe().round())

DEP_DELAY  DEP_DELAY_NEW  DEP_DEL15  DEP_DELAY_GROUP  CARRIER_DELAY  \
count  6551777.0      6551777.0  6551777.0        6551777.0      6551777.0   
mean        13.0           16.0        0.0              0.0            6.0   
std         53.0           52.0        0.0              2.0           36.0   
min        -96.0            0.0        0.0             -2.0            0.0   
25%         -5.0            0.0        0.0             -1.0            0.0   
50%         -2.0            0.0        0.0             -1.0            0.0   
75%         10.0           10.0        0.0              0.0            0.0   
max       3433.0         3433.0        1.0             12.0         3423.0   

       WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  LATE_AIRCRAFT_DELAY  
count      6551777.0  6551777.0       6551777.0            6551777.0  
mean             1.0        2.0             0.0                  5.0  
std             14.0       14.0             2.0                 27.0  
min              0.0        0.0             0.0                  0.0  
25%              0.0        0.0             0.0                  0.0  
50%              0.0        0.0             0.0                  0.0  
75%              0.0        0.0             0.0                  0.0  
max           2363.0     1740.0          1245.0               2302.0

|                         Feature                           |                 Data Quality Requirement                        |
|-----------------------------------------------------------|-----------------------------------------------------------------|
| QUARTER                                                   | Values must range from 1 to 4, by 1 step                        |
|                                                           | Value is required; If Nan, delete row                           |
|                                                           | Value type is int                                               |
| MONTH                                                     | Values must range from 1 to 12, by 1 step                       |
|                                                           | Value is required; If Nan, delete row                           |
|                                                           | Value type is int                                               |
| DAY_OF_MONTH                                              | Values must range from 1 to 28 for month (2)                    |
|                                                           | Values must range from 1 to 30 for month (4, 6, 9, 11)          |
|                                                           | Values must range from 1 to 31 for month (1, 3, 5, 7, 8, 10, 12)|
|                                                           | Value step by 1                                                 |
|                                                           | Value is required; If Nan, delete row                           |
|                                                           | Value type is int                                               |
| DAY_OF_WEEK                                               | Values must range from 1 to 7, by 1 step                        |
|                                                           | Value is required; If Nan, delete row                           |
|                                                           | Value type is int                                               |
| OP_UNIQUE_CARRIER                                         | Value is required; If Nan, delete row                           |
|                                                           | Value type is str                                               |
| OP_CARRIER_AIRLINE_ID                                     | Value is required; If Nan, delete row                           |
|                                                           | Value type is int                                               |
| TAIL_NUM                                                  | Value is required; If Nan, delete row                           |
|                                                           | Value type is str                                               |
| OP_CARRIER_FL_NUM                                         | Value is required; If Nan, delete row                           |
|                                                           | Value type is int                                               |
| ORIGIN_AIRPORT_ID                                         | Value is required; If Nan, delete row                           |
|                                                           | Value type is int                                               |
| ORIGIN_CITY_MARKET_ID                                     | Value is required; If Nan, delete row                           |
|                                                           | Value type is int                                               |
| ORIGIN_STATE_FIPS                                         | Value is required; If Nan, delete row                           |
|                                                           | Value type is int                                               |
| ORIGIN_WAC                                                | Value is required; If Nan, delete row                           |
|                                                           | Value type is int                                               |
| DEST_AIRPORT_ID                                           | Value is required; If Nan, delete row                           |
|                                                           | Value type is int                                               |
| DEST_CITY_MARKET_ID                                       | Value is required; If Nan, delete row                           |
|                                                           | Value type is int                                               |
| DEST_STATE_FIPS                                           | Value is required; If Nan, delete row                           |
|                                                           | Value type is int                                               |
| DEST_WAC                                                  | Value is required; If Nan, delete row                           |
|                                                           | Value type is int                                               |
| CRS_DEP_TIME                                              | Values must range from 1 to 2400, by 1 step                     |
|                                                           | Values with left 2 digits higher than 24 or less than 0 are invalid  |
|                                                           | Values with right 2 digits higher than 59 or less than 0 are invalid |
|                                                           | Value is required; If Nan, delete row                           |
|                                                           | Value type is int                                               |
| DEP_TIME_BLK                                              | Value should have the format "HHMM-HHMM" where:                 |
|                                                           | HH is equal on both sides of separator "-",                     |
|                                                           | "HH" on both sides should range from 0 to 23,                   |
|                                                           | "MM" on the left side should be 0,                              |
|                                                           | "MM" on the right side should be 59                             |
|                                                           | if "MM" on the left side is 00, the right side must be "0559"   |
|                                                           | Value is required; If Nan, delete row                           |
|                                                           | Value type is str                                               |
| CRS_ARR_TIME                                              | Values must range from 1 to 2400, by 1 step                     |
|                                                           | Values with left 2 digits higher than 24 or less than 0 are invalid  |
|                                                           | Values with right 2 digits higher than 59 or less than 0 are invalid |
|                                                           | Value is required; If Nan, delete row                           |
|                                                           | Value type is int                                               |
| ARR_TIME_BLK                                              | Value should have the format "HHMM-HHMM" where:                 |
|                                                           | HH is equal on both sides of separator "-",                     |
|                                                           | "HH" on both sides should range from 0 to 23,                   |
|                                                           | "MM" on the left side should be 0,                              |
|                                                           | "MM" on the right side should be 59                             |
|                                                           | if "MM" on the left side is 00, the right side must be "0559"   |
|                                                           | Value is required; If Nan, delete row                           |
|                                                           | Value type is str                                               |
| CRS_ELAPSED_TIME                                          | Value is required; If Nan, delete row                           |
|                                                           | Negative values are invalid                                     |
|                                                           | Value type is int                                               |
| DISTANCE                                                  | Value is required; If Nan, delete row                           |
|                                                           | Negative values are invalid                                     |
|                                                           | Value type is int                                               |
| DISTANCE_GROUP                                            | Values must range from 1 to 11, by 1 step                       |
|                                                           | Negative values are invalid                                     |
|                                                           | Value is required; If Nan, delete row                           |
|                                                           | Value type is int                                               |
| FLIGHTS                                                   | Negative values are invalid                                     |
|                                                           | Value is required; If Nan, delete row                           |
|                                                           | Value type is int                                               |

In [18]:
# We start by viewing the descriptiuon statstics of each features columns
features = [
    "QUARTER",
    "MONTH",
    "DAY_OF_MONTH",
    "DAY_OF_WEEK",
    
    "OP_UNIQUE_CARRIER",
    "OP_CARRIER_AIRLINE_ID",
    "TAIL_NUM",
    "OP_CARRIER_FL_NUM",
    
    "ORIGIN_AIRPORT_ID",
    "ORIGIN_CITY_MARKET_ID",
    "ORIGIN_STATE_FIPS",
    "ORIGIN_WAC",
    "DEST_AIRPORT_ID",
    "DEST_CITY_MARKET_ID",
    "DEST_STATE_FIPS",
    "DEST_WAC",
    
    "CRS_DEP_TIME",
    "DEP_TIME_BLK",
    "CRS_ARR_TIME",
    "ARR_TIME_BLK",
    "CRS_ELAPSED_TIME",
    "DISTANCE",
    "DISTANCE_GROUP",
    "FLIGHTS",
]

display(prosessed_df[features].describe().round())

QUARTER      MONTH  DAY_OF_MONTH  DAY_OF_WEEK  OP_CARRIER_AIRLINE_ID  \
count  6551777.0  6551777.0     6551777.0    6551777.0              6551777.0   
mean         3.0        7.0          16.0          4.0                19965.0   
std          1.0        3.0           9.0          2.0                  374.0   
min          1.0        1.0           1.0          1.0                19393.0   
25%          2.0        4.0           8.0          2.0                19790.0   
50%          3.0        7.0          16.0          4.0                19930.0   
75%          3.0        9.0          23.0          6.0                20363.0   
max          4.0       12.0          31.0          7.0                20452.0   

       OP_CARRIER_FL_NUM  ORIGIN_AIRPORT_ID  ORIGIN_CITY_MARKET_ID  \
count          6551777.0          6551777.0              6551777.0   
mean              2403.0            12661.0                31746.0   
std               1674.0             1528.0                 1315.0   
min                  1.0            10135.0                30070.0   
25%               1041.0            11292.0                30647.0   
50%               2103.0            12889.0                31453.0   
75%               3550.0            14027.0                32467.0   
max               9562.0            16869.0                36101.0   

       ORIGIN_STATE_FIPS  ORIGIN_WAC  DEST_AIRPORT_ID  DEST_CITY_MARKET_ID  \
count          6551777.0   6551777.0        6551777.0            6551777.0   
mean                27.0        55.0          12661.0              31746.0   
std                 17.0        27.0           1528.0               1315.0   
min                  1.0         1.0          10135.0              30070.0   
25%                 12.0        34.0          11292.0              30666.0   
50%                 26.0        45.0          12889.0              31453.0   
75%                 42.0        82.0          14027.0              32467.0   
max                 78.0        93.0          16869.0              36101.0   

       DEST_STATE_FIPS   DEST_WAC  CRS_DEP_TIME  CRS_ARR_TIME  \
count        6551777.0  6551777.0     6551777.0     6551777.0   
mean              27.0       55.0        1328.0        1486.0   
std               17.0       27.0         492.0         519.0   
min                1.0        1.0           1.0           1.0   
25%               12.0       34.0         910.0        1101.0   
50%               26.0       45.0        1318.0        1512.0   
75%               42.0       82.0        1735.0        1921.0   
max               78.0       93.0        2359.0        2400.0   

       CRS_ELAPSED_TIME   DISTANCE  DISTANCE_GROUP    FLIGHTS  
count         6551777.0  6551777.0       6551777.0  6551777.0  
mean              143.0      817.0             4.0        1.0  
std                73.0      597.0             2.0        0.0  
min               -85.0       31.0             1.0        1.0  
25%                90.0      386.0             2.0        1.0  
50%               125.0      666.0             3.0        1.0  
75%               174.0     1052.0             5.0        1.0  
max               690.0     5095.0            11.0        1.0

In [19]:
## According to the established data quality requirements above, all of the features in the table above should have values in their column data cells.
# Value is required; If Nan, delete row

# Count rows with NaN in features columns
count_nan_rows = prosessed_df[prosessed_df[features].isnull().any(axis=1)].shape[0]

# Print count of rows with NaN
print(f"Count of rows with NaN in features: {count_nan_rows}")

# Drop rows with NaN in columns A or B
prosessed_cleaned = prosessed_df.dropna(subset=features)

# Print count of remaining rows
remaining_count = prosessed_cleaned.shape[0]
print(f"Count of remaining rows: {remaining_count}")

Count of rows with NaN in features: 0
Count of remaining rows: 6551777


In [20]:
## According to the established data quality requirements above, the features column below should be of type int
# Here we make sure these column are in dtype int or convert them to int

int_features = [
    "QUARTER",
    "MONTH",
    "DAY_OF_MONTH",
    "DAY_OF_WEEK",
    "OP_CARRIER_AIRLINE_ID",
    "OP_CARRIER_FL_NUM",
    "ORIGIN_AIRPORT_ID",
    "ORIGIN_CITY_MARKET_ID",
    "ORIGIN_STATE_FIPS",
    "ORIGIN_WAC",
    "DEST_AIRPORT_ID",
    "DEST_CITY_MARKET_ID",
    "DEST_STATE_FIPS",
    "DEST_WAC",
    "CRS_DEP_TIME",
    "CRS_ARR_TIME",
    "CRS_ELAPSED_TIME",
    "DISTANCE",
    "DISTANCE_GROUP",
    "FLIGHTS"
]

for column in int_features:
    # Convert to int
    print(f"Converting to {column} dtype int")
    print(f"Before {prosessed_df[column].dtype = }")
    prosessed_df[column] = prosessed_df[column].astype(int)
    print(f"After {prosessed_df[column].dtype = }")
    print()


Converting to QUARTER dtype int
Before prosessed_df[column].dtype = dtype('int64')
After prosessed_df[column].dtype = dtype('int64')

Converting to MONTH dtype int
Before prosessed_df[column].dtype = dtype('int64')
After prosessed_df[column].dtype = dtype('int64')

Converting to DAY_OF_MONTH dtype int
Before prosessed_df[column].dtype = dtype('int64')
After prosessed_df[column].dtype = dtype('int64')

Converting to DAY_OF_WEEK dtype int
Before prosessed_df[column].dtype = dtype('int64')
After prosessed_df[column].dtype = dtype('int64')

Converting to OP_CARRIER_AIRLINE_ID dtype int
Before prosessed_df[column].dtype = dtype('int64')
After prosessed_df[column].dtype = dtype('int64')

Converting to OP_CARRIER_FL_NUM dtype int
Before prosessed_df[column].dtype = dtype('int64')
After prosessed_df[column].dtype = dtype('int64')

Converting to ORIGIN_AIRPORT_ID dtype int
Before prosessed_df[column].dtype = dtype('int64')
After prosessed_df[column].dtype = dtype('int64')

Converting to ORIGIN_

In [21]:
## According to the established data quality requirements above, the features column below should be of type str
# Here we make sure these column are in dtype str/Object or convert them to str/Object

str_features = [
    "OP_UNIQUE_CARRIER",
    "TAIL_NUM",
    "DEP_TIME_BLK",
    "ARR_TIME_BLK"
]

for column in str_features:
    # Convert to str
    print(f"Converting to {column} dtype int")
    print(f"Before {prosessed_df[column].dtype = }")
    prosessed_df[column] = prosessed_df[column].astype(str)
    print(f"After {prosessed_df[column].dtype = }")
    print()

Converting to OP_UNIQUE_CARRIER dtype int
Before prosessed_df[column].dtype = dtype('O')
After prosessed_df[column].dtype = dtype('O')

Converting to TAIL_NUM dtype int
Before prosessed_df[column].dtype = dtype('O')
After prosessed_df[column].dtype = dtype('O')

Converting to DEP_TIME_BLK dtype int
Before prosessed_df[column].dtype = dtype('O')
After prosessed_df[column].dtype = dtype('O')

Converting to ARR_TIME_BLK dtype int
Before prosessed_df[column].dtype = dtype('O')
After prosessed_df[column].dtype = dtype('O')



In [22]:
## QUARTER
# This feature should only have values from 1 to 4 according to the data requirement
# Drop any rows that does not comply with thise requirement

# Drop rows where values in column 'QUARTER' are not in [1, 2, 3, 4]
print("Dropping rows where values in column 'QUARTER' are not in [1, 2, 3, 4]")
non_valid_count = (~prosessed_df['QUARTER'].isin([1, 2, 3, 4])).sum()
print(f"{non_valid_count = }")
prosessed_df = prosessed_df[prosessed_df['QUARTER'].isin([1, 2, 3, 4])]
print(f"After {len(prosessed_df) = }")

Dropping rows where values in column 'QUARTER' are not in [1, 2, 3, 4]
non_valid_count = np.int64(0)
After len(prosessed_df) = 6551777


In [23]:
## MONTH
# This feature should only have values from 1 to 12 according to the data requirement
# Drop any rows that does not comply with thise requirement

# Drop rows where values in column 'Month' are not in [1, 2, 3, 4]
print("Dropping rows where values in column 'MONTH' are not in [1, 2, 3, 4, ..., 10, 11, 12]")
valid_values = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
non_valid_count = (~prosessed_df['MONTH'].isin(valid_values)).sum()
print(f"{non_valid_count = }")
prosessed_df = prosessed_df[prosessed_df['MONTH'].isin(valid_values)]
print(f"After {len(prosessed_df) = }")

Dropping rows where values in column 'MONTH' are not in [1, 2, 3, 4, ..., 10, 11, 12]
non_valid_count = np.int64(0)
After len(prosessed_df) = 6551777


In [24]:
## DAY_OF_MONTH
# This feature column should have the following values for both of the years 2022 and 2023
#Values must range from 1 to 28 for month (2)
#Values must range from 1 to 30 for month (4, 6, 9, 11)
#Values must range from 1 to 31 for month (1, 3, 5, 7, 8, 10, 12)
#Value step by 1
# Thus we need to combine our search and control with the MONTH feature

# Days in each month (not accounting for leap years)
days_in_month = {
    1: 31,
    2: 28,  # Adjust february for leap years if necessary for future dataset processing
    3: 31,
    4: 30,
    5: 31,
    6: 30,
    7: 31,
    8: 31,
    9: 30,
    10: 31,
    11: 30,
    12: 31,
}

# Function to check validity of the day
def is_valid_day(row):
    month = row['MONTH']
    day = row['DAY_OF_MONTH']
    max_days = days_in_month.get(month, 0)
    return 1 <= day <= max_days

# Find valid_rows
valid_rows = prosessed_df.apply(is_valid_day, axis=1)

# Count invalid rows
invalid_count = prosessed_df[~valid_rows].shape[0]
print(f"Count of invalid rows: {invalid_count}")

# Remove invalid rows
prosessed_df = prosessed_df[valid_rows]
print(f"After {len(prosessed_df) = }")

Count of invalid rows: 0
After len(prosessed_df) = 6551777


In [25]:
## DAY_OF_WEEK
# This feature should only have values from 1 to 7 according to the data requirement
# Drop any rows that does not comply with thise requirement

# Drop rows where values in column 'DAY_OF_WEEK' are not in [1, 2, 3, 4, 5, 6, 7]
print("Dropping rows where values in column 'DAY_OF_WEEK' are not in [1, 2, 3, 4, 5, 6, 7]")
non_valid_count = (~prosessed_df['DAY_OF_WEEK'].isin([1, 2, 3, 4, 5, 6, 7])).sum()
print(f"Count of invalid rows: {non_valid_count}")
prosessed_df = prosessed_df[prosessed_df['DAY_OF_WEEK'].isin([1, 2, 3, 4, 5, 6, 7])]
print(f"After {len(prosessed_df) = }")

Dropping rows where values in column 'DAY_OF_WEEK' are not in [1, 2, 3, 4, 5, 6, 7]
Count of invalid rows: 0
After len(prosessed_df) = 6551777


In [26]:
## CRS_DEP_TIME, CRS_ARR_TIME
# This 2 columns feature above should have the following quality requirements
# Values must range from 1 to 2400, by 1 step
# Values with left 2 digits higher than 24 or less than 0 are invalid
# Values with right 2 digits higher than 59 or less than 0 are invalid

for column in ["CRS_DEP_TIME", "CRS_ARR_TIME"]:
    print(f"Removing invalid values from {column}")
    
    # Convert column to string to handle leading zeros
    prosessed_df[column] = prosessed_df[column].astype(str).str.zfill(4)
    
    # Check for invalid rows for hour
    invalid_rows_hours = prosessed_df[
        (prosessed_df[column].str[:2].astype(int) > 24) | (prosessed_df[column].str[:2].astype(int) < 0)
    ]
    
    # Check for invalid rows for minutes
    invalid_rows_minutes = prosessed_df[
        (prosessed_df[column].str[2:].astype(int) > 59) | (prosessed_df[column].str[2:].astype(int) < 0)
    ]

    # Check for invalid rows with 0 as value
    invalid_rows_zero = prosessed_df[
        prosessed_df[column].str[:4].astype(int) <= 0
    ]
    
    # Print the count of invalid rows
    invalid_count = len(invalid_rows_hours) + len(invalid_rows_minutes) + len(invalid_rows_zero)
    print(f'Number of invalid rows: {invalid_count}')
    
    # Remove invalid rows from the DataFrame
    prosessed_df = prosessed_df.drop(invalid_rows_hours.index)
    prosessed_df = prosessed_df.drop(invalid_rows_minutes.index)
    prosessed_df = prosessed_df.drop(invalid_rows_zero.index)
    
    # Convert column back to integer
    prosessed_df[column] = prosessed_df[column].astype(int)
    
    # Print the remaining row count after cleaning
    print(f"After {len(prosessed_df) = }")
    print()

Removing invalid values from CRS_DEP_TIME
Number of invalid rows: 0
After len(prosessed_df) = 6551777

Removing invalid values from CRS_ARR_TIME
Number of invalid rows: 0
After len(prosessed_df) = 6551777



In [27]:
## DEP_TIME_BLK, ARR_TIME_BLK
# This 2 columns feature above should have the following quality requirements
# Value should have the foramt "HHMM-HHMM" where:
# HH is equal on both sides of separator "-", 
# "HH" on both sides should range from 0 to 23, 
# "MM" on the left side should be 0, 
# "MM" on the right side should be 59
# if "MM" on the left side is 00, the right side must be "0559"


# Define a function to validate the format
def is_valid_time_blk(time_blk):
    try:
        # Split the time block
        left, right = time_blk.split('-')

        if left == "0001" and right == "0559":
            return True
        elif left == "0001" and right != "0559":
            return False
        else:
            # Extract hours and minutes
            hh_left, mm_left = int(left[:2]), int(left[2:])
            hh_right, mm_right = int(right[:2]), int(right[2:])
            
            # Validate the conditions
            return (
                hh_left == hh_right and
                0 <= hh_left <= 23 and
                0 <= hh_right <= 23 and
                mm_left == 0 and
                mm_right == 59
            )
    except Exception as e:
        return False


for column in ["DEP_TIME_BLK", "ARR_TIME_BLK"]:
    print(f"Removing invalid values from {column}")

    # Apply the validation function
    valid_mask = prosessed_df[column].apply(is_valid_time_blk)
    
    # Count invalid rows
    invalid_count = (~valid_mask).sum()
    print(f'Count of invalid rows: {invalid_count}')
    
    # Remove invalid rows from the DataFrame
    prosessed_df = prosessed_df[valid_mask].reset_index(drop=True)

    # Print the remaining row count after cleaning
    print(f"After {len(prosessed_df) = }")
    print()

Removing invalid values from DEP_TIME_BLK
Count of invalid rows: 0
After len(prosessed_df) = 6551777

Removing invalid values from ARR_TIME_BLK
Count of invalid rows: 0
After len(prosessed_df) = 6551777



In [28]:
## CRS_ELAPSED_TIME, DISTANCE, DISTANCE_GROUP, FLIGHTS
# This 4 columns feature above should not have any negative values

# Find rows with any negative values in the specified columns
invalid_rows = prosessed_df[["CRS_ELAPSED_TIME", "DISTANCE", "DISTANCE_GROUP", "FLIGHTS"]].lt(0).any(axis=1)

# Count of invalid rows
count_invalid = invalid_rows.sum()
print(f"Count of invalid rows: {count_invalid}")

# Drop invalid rows from the DataFrame
prosessed_df = prosessed_df.loc[~invalid_rows].reset_index(drop=True)

# Print the remaining row count after cleaning
print(f"After {len(prosessed_df) = }")
print()

Count of invalid rows: 4
After len(prosessed_df) = 6551773



In [29]:
## DISTANCE_GROUP
# The values of column DISTANCE_GROUP must range from 1 to 11, by 1 step

# Drop rows where values in column 'DISTANCE_GROUP' are not in [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
print("Dropping rows where values in column 'DISTANCE_GROUP' are not in [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]")
non_valid_count = (~prosessed_df['DISTANCE_GROUP'].isin([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11])).sum()
print(f"Count of invalid rows: {non_valid_count}")
prosessed_df = prosessed_df[prosessed_df['DISTANCE_GROUP'].isin([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11])]
print(f"After {len(prosessed_df) = }")

Dropping rows where values in column 'DISTANCE_GROUP' are not in [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
Count of invalid rows: 0
After len(prosessed_df) = 6551773


## Sum up
- The table below sums up the preprocessing, cleaning and quality control by viewing the number of invalid rows found in each column
- We started off with 6 729 125 records in dataset

|                         Feature                           |               Number of invalid rows found                      |
|-----------------------------------------------------------|-----------------------------------------------------------------|
| QUARTER                                                   | 0                                                               |
| MONTH                                                     | 0                                                               |
| DAY_OF_MONTH                                              | 0                                                               |
| DAY_OF_WEEK                                               | 0                                                               |
| OP_UNIQUE_CARRIER                                         | 0                                                               |
| OP_CARRIER_AIRLINE_ID                                     | 0                                                               |
| TAIL_NUM                                                  | 0                                                               |
| OP_CARRIER_FL_NUM                                         | 0                                                               |
| ORIGIN_AIRPORT_ID                                         | 0                                                               |
| ORIGIN_CITY_MARKET_ID                                     | 0                                                               |
| ORIGIN_STATE_FIPS                                         | 0                                                               |
| ORIGIN_WAC                                                | 0                                                               |
| DEST_AIRPORT_ID                                           | 0                                                               |
| DEST_CITY_MARKET_ID                                       | 0                                                               |
| DEST_STATE_FIPS                                           | 0                                                               |
| DEST_WAC                                                  | 0                                                               |
| CRS_DEP_TIME                                              | 0                                                               |
| DEP_TIME_BLK                                              | 0                                                               |
| CRS_ARR_TIME                                              | 0                                                               |
| ARR_TIME_BLK                                              | 0                                                               |
| CRS_ELAPSED_TIME                                          | (-4)                                                            |
| DISTANCE                                                  | 0                                                               |
| DISTANCE_GROUP                                            | 0                                                               |
| FLIGHTS                                                   | 0                                                               |
| DEP_DELAY                                                 | (-177 348)                                                       |
| DEP_DELAY_NEW                                             | 0                                                               |
| DEP_DEL15                                                 | 0                                                               |
| DEP_DELAY_GROUP                                           | 0                                                               |
| CARRIER_DELAY                                             | 0                                                               |
| WEATHER_DELAY                                             | 0                                                               |
| NAS_DELAY                                                 | 0                                                               |
| SECURITY_DELAY                                            | 0                                                               |
| LATE_AIRCRAFT_DELAY                                       | 0                                                               |

- The remaining records count is 6 551 773